forch/faucet_state_collector.py

```diff
@@ -569,24 +571,72 @@ def _fill_path_to_root(self, switch_name, switch_map):
         """populate path to root for switch_state"""
         switch_map["root_path"] = self.get_switch_egress_path(switch_name)
 
-    def _fill_dva_states(self, switch_name, port_id, port_map):
-        dp_obj = self.faucet_config.get(DPS_CFG, {}).get(switch_name)
-        if not dp_obj:
+    def _fill_vlan_behavior(self, switch_name, switch_map, metrics=None):
+        dp_config = self.faucet_config.get(DPS_CFG, {}).get(switch_name)
+        if not dp_config:
             LOGGER.warning('Switch not defined in dps config: %s', switch_name)
             return
 
-        port_obj = dp_obj.interfaces.get(port_id)
-        if not port_obj:
+        vlans_map = switch_map.setdefault('vlans', {})
+        for vid, vlan_config in dp_config.vlans.items():
+            acls_map = vlans_map.setdefault(int(vid), {}).setdefault('acls', {})
+            if metrics and 'flow_packet_count_vlan_acl' in metrics:
+                samples = metrics['flow_packet_count_vlan_acl'].samples
+            else:
+                samples = None
+            self._fill_rules_behavior(switch_name, acls_map, vlan_config.acls_in, samples)
+
+    def _fill_port_behavior(self, switch_name, port_id, port_map, metrics=None):
+        dp_config = self.faucet_config.get(DPS_CFG, {}).get(switch_name)
+        if not dp_config:
+            LOGGER.warning('Switch not defined in dps config: %s', switch_name)
+            return
+
+        port_config = dp_config.ports.get(port_id)
+        if not port_config:
             LOGGER.warning('Port not defined in dps config: %s, %s', switch_name, port_id)
             return
 
-        native_vlan = port_obj.get('native_vlan')
-        acls_in = port_obj.get('acls_in')
+        if port_config.native_vlan:
+            port_map['vlan'] = int(port_config.native_vlan.vid)
+        if port_config.acls_in:
+            acls_map = port_map.setdefault('acls', {})
+            if metrics and 'flow_packet_count_port_acl' in metrics:
+                samples = metrics['flow_packet_count_port_acl'].samples
+            else:
+                samples = None
+            self._fill_rules_behavior(
+                switch_name, acls_map, port_config.acls_in, samples, port_id)
+
+    def _fill_rules_behavior(self, switch_name, acls_map, acls_config,
+                             metric_samples=None, port_id=None):
+        for acl_config in acls_config:
+            rules_map_list = acls_map.setdefault(str(acl_config._id), {}).setdefault('rules', [])
+            for rule_config in acl_config.rules:
+                cookie = str(rule_config.get('cookie'))
+                rule_map = {
+                    'cookie': cookie,
+                    'description': rule_config.get('description')
+                }
+                rules_map_list.append(rule_map)
+
+                if not metric_samples:
+                    continue
+
+                if not cookie:
+                    LOGGER.warning('ACL %s does not have a cookie', acl_config._id)
```

Comment:
I don't think we can reasonably require all ACLs to have a cookie on them, so I'm not sure this should be a warning

Reply:
OK, removed. I think we can say that by assigning an ACL rule a cookie, we want the stats of this rule to be monitored. Otherwise we do not care about the stats but only the rule configuration.

Reply:
That is fair for now, but I'm pretty sure that we'll just want to enable it for everything -- I don't see why not (it's not computationally expensive).  I would consider a mechanism where cookies are automatically assigned -- seems much cleaner overall to me (push the cookie abstraction down under the hood, rather than expose it).

Reply:
Raising an exception now since we expect faucetizer always assign cookie to each rule.